libcaf_net/src/ip.cpp

```diff
@@ -113,6 +146,114 @@ std::string hostname() {
   return buf;
 }
 
+#ifdef CAF_WINDOWS
+
+std::vector<ip_address> local_addresses(string_view host) {
+  using adapters_ptr = std::unique_ptr<IP_ADAPTER_ADDRESSES, void (*)(void*)>;
+  using wbuf_ptr = std::unique_ptr<IP_ADAPTER_ADDRESSES, void (*)(void*)>;
+  if (!host.empty() && host.compare(v4_any_addr) == 0)
+    return {ip_address{make_ipv4_address(0, 0, 0, 0)}};
+  if (!host.empty() && host.compare(v6_any_addr) == 0)
+    return {ip_address{}};
+  ULONG len = 0;
+  auto err = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, nullptr,
+                                  nullptr, &len);
+  if (err != ERROR_BUFFER_OVERFLOW) {
+    CAF_LOG_ERROR("failed to get adapter addresses buffer length");
+    return {};
+  }
+  auto adapters = adapters_ptr{reinterpret_cast<IP_ADAPTER_ADDRESSES*>(
+                                 ::new uint8_t[len]),
+                               free};
+  if (!adapters) {
+    CAF_LOG_ERROR("malloc failed");
+    return {};
+  }
+  // TODO: The API example propopses to try three times.
+  err = GetAdaptersAddresses(AF_UNSPEC, GAA_FLAG_INCLUDE_PREFIX, nullptr,
+                             adapters.get(), &len);
+  if (err != ERROR_SUCCESS) {
+    CAF_LOG_ERROR("failed to get adapter addresses");
+    return {};
+  }
+  char ip_buf[INET6_ADDRSTRLEN];
+  char name_buf[HOST_NAME_MAX];
+  std::vector<std::pair<std::string, ip_address>> interfaces;
```

Comment:
Why are we filling up this vector instead of doing the work right there in the for loop?

I think if we move the loop into a template function, then we can safe a lot of unnecessary work and also make the code more modular:

```
template <class F>
void for_each_adapter(IP_ADAPTER_ADDRESSES* first, F f) {
  for (; first != nullptr; first = first->Next) {
    // ...
  }
}

// Call side.
if (host.empty()) {
  auto f = [&](string_view, ip_address addr) {
    results.push_back(addr);
  };
  for_each_adapter(adapters.get(), f);
} else if ...
```